File: typescript/examples/agents/applied_agents/daily-digest-agent.ts
/**
 * Daily Digest Agent — direct API tools, no MCP middleware.
 *
 * Builds a daily digest by:
 * 1. Fetching the top Hacker News posts (free, no auth).
 * 2. Searching the web for restaurant recommendations via Jina AI's search API.
 * 3. Writing the result as a structured Notion page with action items.
 *
 * This example shows the opperai Agent loop driving plain `tool({ ... })`
 * functions against three small public APIs — no MCP server, no third-party
 * tool broker.
 *
 * Prerequisites:
 *   - OPPER_API_KEY in .env
 *   - JINA_API_KEY in .env (https://jina.ai/?sui=apikey)
 *   - NOTION_TOKEN in .env (notion.so/profile/integrations, "Internal" integration)
 *   - NOTION_PARENT_PAGE_ID in .env — a Notion page ID where the digest will be
 *     created as a child. Share that page with your integration first
 *     (Notion → page → Share → add the integration).
 *
 * Run with:
 *   cd typescript
 *   node --env-file=../.env node_modules/.bin/tsx \
 *     examples/agents/applied_agents/daily-digest-agent.ts
 */

import { z } from "zod";
import { Agent, tool } from "../../../src/index.js";
import type { Hooks } from "../../../src/index.js";

// ---------------------------------------------------------------------------
// Schemas
// ---------------------------------------------------------------------------

const NewsItemSchema = z.object({
  title: z.string().describe("News title"),
  url: z.string().nullable().optional().describe("Link to the article"),
  points: z.number().nullable().optional().describe("Number of points / upvotes"),
  summary: z.string().describe("Brief summary or commentary on the story"),
});

const RestaurantPickSchema = z.object({
  name: z.string().describe("Restaurant name"),
  location: z.string().describe("City and neighbourhood"),
  cuisine: z.string().describe("Type of cuisine"),
  reason: z.string().describe("Why it is worth a visit"),
  url: z.string().nullable().optional().describe("Source URL with more info"),
});

const ActionItemSchema = z.object({
  action: z.string().describe("The action to take"),
  source: z.enum(["news", "restaurant"]),
  link: z.string().nullable().optional().describe("Related link if available"),
});

const DailyDigestSchema = z.object({
  date: z.string().describe("Date of the digest, ISO 8601 (YYYY-MM-DD)"),
  news: z.array(NewsItemSchema).describe("Top news items from Hacker News"),
  restaurants: z.array(RestaurantPickSchema).describe("Restaurants worth trying"),
  actions: z.array(ActionItemSchema).describe("Action items derived from the digest"),
  notion_page_url: z.string().describe("URL of the created Notion page"),
});

// ---------------------------------------------------------------------------
// Tools
// ---------------------------------------------------------------------------

const HN_BASE = "https://hacker-news.firebaseio.com/v0";

const fetchHnTopStories = tool({
  name: "fetch_hn_top_stories",
  description:
    "Fetch the top stories currently on Hacker News. Returns id, title, url, score (upvotes), author, and comment count.",
  parameters: z.object({
    limit: z.number().int().min(1).max(30).default(10).describe("Number of stories to fetch"),
  }),
  execute: async ({ limit }) => {
    const idsResp = await fetch(`${HN_BASE}/topstories.json`);
    if (!idsResp.ok) throw new Error(`HN topstories: HTTP ${idsResp.status}`);
    const ids = (await idsResp.json()) as number[];
    const picked = ids.slice(0, limit);

    const items = await Promise.all(
      picked.map(async (id) => {
        const r = await fetch(`${HN_BASE}/item/${id}.json`);
        if (!r.ok) throw new Error(`HN item ${id}: HTTP ${r.status}`);
        return (await r.json()) as Record<string, unknown>;
      }),
    );

    return items.map((it) => ({
      id: it.id,
      title: it.title,
      url: it.url,
      score: it.score,
      by: it.by,
      descendants: it.descendants,
    }));
  },
});

const searchWeb = tool({
  name: "search_web",
  description:
    "Search the web via Jina AI's search API. Pass a natural-language query (e.g. 'best ramen restaurants in Stockholm 2026'). Returns {title, url, description} results.",
  parameters: z.object({
    query: z.string().describe("Natural-language search query"),
    max_results: z.number().int().min(1).max(10).default(5),
  }),
  execute: async ({ query, max_results }) => {
    const apiKey = process.env.JINA_API_KEY;
    if (!apiKey) throw new Error("JINA_API_KEY not set");

    const resp = await fetch(`https://s.jina.ai/${encodeURIComponent(query)}`, {
      headers: {
        Authorization: `Bearer ${apiKey}`,
        Accept: "application/json",
        "X-Respond-With": "no-content",
      },
    });
    if (!resp.ok) throw new Error(`Jina search: HTTP ${resp.status}`);
    const json = (await resp.json()) as { data?: Array<Record<string, unknown>> };
    const data = json.data ?? [];

    return data.slice(0, max_results).map((item) => ({
      title: item.title,
      url: item.url,
      description: item.description ?? item.content ?? "",
    }));
  },
});

const createNotionPage = tool({
  name: "create_notion_page",
  description:
    "Create a Notion page as a child of NOTION_PARENT_PAGE_ID. `sections` is a list of {heading, body}. Each section is rendered as a heading_2 + paragraph block. Returns the URL of the created page.",
  parameters: z.object({
    title: z.string().describe("Page title"),
    sections: z
      .array(
        z.object({
          heading: z.string(),
          body: z.string(),
        }),
      )
      .describe("Sections to render as heading + paragraph blocks"),
  }),
  execute: async ({ title, sections }) => {
    const token = process.env.NOTION_TOKEN;
    const parentId = process.env.NOTION_PARENT_PAGE_ID;
    if (!token) throw new Error("NOTION_TOKEN not set");
    if (!parentId) throw new Error("NOTION_PARENT_PAGE_ID not set");

    const children: Array<Record<string, unknown>> = [];
    for (const section of sections) {
      const heading = (section.heading ?? "").trim();
      const body = (section.body ?? "").trim();
      if (heading) {
        children.push({
          object: "block",
          type: "heading_2",
          heading_2: {
            rich_text: [{ type: "text", text: { content: heading } }],
          },
        });
      }
      if (body) {
        // Notion caps text blocks at ~2000 chars; chunk conservatively.
        const chunks: string[] = [];
        for (let i = 0; i < body.length; i += 1900) chunks.push(body.slice(i, i + 1900));
        for (const chunk of chunks.length ? chunks : [""]) {
          children.push({
            object: "block",
            type: "paragraph",
            paragraph: {
              rich_text: [{ type: "text", text: { content: chunk } }],
            },
          });
        }
      }
    }

    const payload = {
      parent: { page_id: parentId },
      properties: {
        title: { title: [{ type: "text", text: { content: title } }] },
      },
      children,
    };

    const resp = await fetch("https://api.notion.com/v1/pages", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${token}`,
        "Notion-Version": "2022-06-28",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
    });
    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`Notion: HTTP ${resp.status} — ${text}`);
    }
    const json = (await resp.json()) as { url?: string };
    return json.url ?? "";
  },
});

// ---------------------------------------------------------------------------
// Agent
// ---------------------------------------------------------------------------

const today = new Date();
const dateStr = today.toISOString().split("T")[0];
const dayName = today.toLocaleDateString("en-US", { weekday: "long" });

const instructions = `
You are a personal productivity assistant that produces a daily digest.

Workflow for ${dateStr} (${dayName}):
1. Fetch the top Hacker News posts (use fetch_hn_top_stories).
   - Pick the most interesting 5-7.
   - Summarize what makes each one notable.

2. Search the web for restaurant recommendations (use search_web).
   - Aim for highly-rated restaurants in Stockholm with recent reviews.
   - Pick 3-5 worth trying this weekend.

3. Derive 2-4 action items from the news and restaurants.

4. Create a Notion page titled "Daily Digest - ${dateStr}" (use
   create_notion_page) with sections:
   - "Top News"
   - "Restaurant Picks"
   - "Actions"
   Render each section's body as concise plain text — one line per item with
   the title and link.

Be concise. Always include source URLs. Iterate through every step; do not
return an empty digest.
`.trim();

const hooks: Hooks = {
  onIterationStart: ({ iteration }) => {
    console.log(`\n--- Iteration ${iteration} ---`);
  },
  onToolStart: ({ name, input }) => {
    const s = JSON.stringify(input);
    console.log(`  -> ${name}(${s.slice(0, 120)}${s.length > 120 ? "..." : ""})`);
  },
  onToolEnd: ({ name, output, error, durationMs }) => {
    if (error) {
      console.log(`  x ${name} failed (${durationMs}ms): ${error}`);
    } else {
      const s = JSON.stringify(output);
      console.log(`  <- ${name} (${durationMs}ms): ${s.slice(0, 150)}${s.length > 150 ? "..." : ""}`);
    }
  },
  onAgentEnd: ({ result, error }) => {
    if (error) {
      console.log(`\nx Agent failed: ${error.message}`);
    } else if (result) {
      console.log(`\nv Agent completed in ${result.meta.iterations} iteration(s)`);
    }
  },
};

const agent = new Agent({
  name: "daily-digest",
  instructions,
  outputSchema: DailyDigestSchema,
  model: "anthropic/claude-sonnet-4-6",
  tools: [fetchHnTopStories, searchWeb, createNotionPage],
  maxIterations: 15,
  hooks,
});

// ---------------------------------------------------------------------------
// Run
// ---------------------------------------------------------------------------

for (const key of ["OPPER_API_KEY", "JINA_API_KEY", "NOTION_TOKEN", "NOTION_PARENT_PAGE_ID"]) {
  if (!process.env[key]) {
    console.error(`Missing required env var: ${key}`);
    console.error("See the comment block at the top of this file for setup steps.");
    process.exit(1);
  }
}

console.log("Daily Digest Agent");
console.log("=".repeat(60));
console.log(`Date: ${dateStr} (${dayName})\n`);
console.log("Running...\n");

try {
  const result = await agent.run(`Create my daily digest for ${dateStr} (${dayName}).`);
  const digest = result.output;
  console.log("=".repeat(60));
  console.log("DAILY DIGEST CREATED");
  console.log("=".repeat(60));
  console.log(`\nDate: ${digest.date}`);
  console.log(`News items: ${digest.news.length}`);
  console.log(`Restaurants: ${digest.restaurants.length}`);
  console.log(`Actions: ${digest.actions.length}`);
  console.log(`\nNotion page: ${digest.notion_page_url}`);
  console.log(`Tokens used: ${result.meta.usage.totalTokens}`);
  console.log(`Iterations: ${result.meta.iterations}`);
  console.log(`Tool calls: ${result.meta.toolCalls.length}`);
} catch (error: unknown) {
  const err = error as Error & { cause?: Error & { body?: unknown } };
  console.error(`\nError: ${err.message}`);
  if (err.cause) {
    console.error(`Cause: ${err.cause.message}`);
    if (err.cause.body) console.error("Body:", JSON.stringify(err.cause.body, null, 2));
  }
  process.exit(1);
}

File: python/examples/agents/applied_agents/daily_digest_agent.py
"""
Daily Digest Agent — direct API tools, no MCP middleware.

Builds a daily digest by:
1. Fetching the top Hacker News posts (free, no auth).
2. Searching the web for restaurant recommendations via Jina AI's search API.
3. Writing the result as a structured Notion page with action items.

This example shows the opperai Agent loop driving plain ``@tool`` functions
against three small public APIs — no MCP server, no third-party tool broker.

Prerequisites:
  - OPPER_API_KEY in .env
  - JINA_API_KEY in .env (https://jina.ai/?sui=apikey)
  - NOTION_TOKEN in .env (notion.so/profile/integrations, "Internal" integration)
  - NOTION_PARENT_PAGE_ID in .env — a Notion page ID where the digest will be
    created as a child. Share that page with your integration first
    (Notion → page → Share → add the integration).

Run with:
  PYTHONPATH=src uv run python examples/agents/applied_agents/daily_digest_agent.py
"""

import asyncio
import json
import os
from datetime import datetime
from typing import Any

import httpx
from pydantic import BaseModel, Field

from opperai.agent import Agent, Hooks, tool

# ---------------------------------------------------------------------------
# Schemas
# ---------------------------------------------------------------------------


class NewsItem(BaseModel):
    title: str = Field(description="News title")
    url: str | None = Field(default=None, description="Link to the article")
    points: int | None = Field(default=None, description="Number of points / upvotes")
    summary: str = Field(description="Brief summary or commentary on the story")


class RestaurantPick(BaseModel):
    name: str = Field(description="Restaurant name")
    location: str = Field(description="City and neighbourhood")
    cuisine: str = Field(description="Type of cuisine")
    reason: str = Field(description="Why it is worth a visit")
    url: str | None = Field(default=None, description="Source URL with more info")


class ActionItem(BaseModel):
    action: str = Field(description="The action to take")
    source: str = Field(description="news or restaurant")
    link: str | None = Field(default=None, description="Related link if available")


class DailyDigest(BaseModel):
    date: str = Field(description="Date of the digest, ISO 8601 (YYYY-MM-DD)")
    news: list[NewsItem] = Field(description="Top news items from Hacker News")
    restaurants: list[RestaurantPick] = Field(description="Restaurants worth trying")
    actions: list[ActionItem] = Field(description="Action items derived from the digest")
    notion_page_url: str = Field(description="URL of the created Notion page")


# ---------------------------------------------------------------------------
# Tools
# ---------------------------------------------------------------------------


HN_BASE = "https://hacker-news.firebaseio.com/v0"


@tool
async def fetch_hn_top_stories(limit: int = 10) -> list[dict[str, Any]]:
    """Fetch the top stories currently on Hacker News.

    Returns a list of stories with id, title, url, score (upvotes), author,
    and descendants (comment count).
    """
    capped = max(1, min(limit, 30))
    async with httpx.AsyncClient(timeout=15) as client:
        ids_resp = await client.get(f"{HN_BASE}/topstories.json")
        ids_resp.raise_for_status()
        story_ids: list[int] = ids_resp.json()[:capped]

        async def fetch_item(sid: int) -> dict[str, Any]:
            r = await client.get(f"{HN_BASE}/item/{sid}.json")
            r.raise_for_status()
            return r.json() or {}

        items = await asyncio.gather(*(fetch_item(sid) for sid in story_ids))

    return [
        {
            "id": it.get("id"),
            "title": it.get("title"),
            "url": it.get("url"),
            "score": it.get("score"),
            "by": it.get("by"),
            "descendants": it.get("descendants"),
        }
        for it in items
    ]


@tool
async def search_web(query: str, max_results: int = 5) -> list[dict[str, Any]]:
    """Search the web with Jina AI's search API.

    Pass a natural-language query (for example
    'best ramen restaurants in Stockholm 2026'). Returns a list of
    {title, url, description} results.
    """
    api_key = os.environ["JINA_API_KEY"]
    cap = max(1, min(max_results, 10))
    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.get(
            f"https://s.jina.ai/{query}",
            headers={
                "Authorization": f"Bearer {api_key}",
                "Accept": "application/json",
                "X-Respond-With": "no-content",
            },
        )
        resp.raise_for_status()
        data = resp.json().get("data", []) or []

    return [
        {
            "title": item.get("title"),
            "url": item.get("url"),
            "description": item.get("description") or item.get("content") or "",
        }
        for item in data[:cap]
    ]


@tool
async def create_notion_page(title: str, sections: list[dict[str, str]]) -> str:
    """Create a Notion page as a child of NOTION_PARENT_PAGE_ID.

    ``sections`` is a list of ``{"heading": str, "body": str}`` entries.
    Each section is rendered as a heading_2 block followed by a paragraph
    block. Returns the URL of the created page.
    """
    token = os.environ["NOTION_TOKEN"]
    parent_id = os.environ["NOTION_PARENT_PAGE_ID"]

    children: list[dict[str, Any]] = []
    for section in sections:
        heading = (section.get("heading") or "").strip()
        body = (section.get("body") or "").strip()
        if heading:
            children.append(
                {
                    "object": "block",
                    "type": "heading_2",
                    "heading_2": {
                        "rich_text": [{"type": "text", "text": {"content": heading}}]
                    },
                }
            )
        if body:
            # Notion caps text blocks at ~2000 chars; split conservatively.
            for chunk in [body[i : i + 1900] for i in range(0, len(body), 1900)] or [""]:
                children.append(
                    {
                        "object": "block",
                        "type": "paragraph",
                        "paragraph": {
                            "rich_text": [{"type": "text", "text": {"content": chunk}}]
                        },
                    }
                )

    payload = {
        "parent": {"page_id": parent_id},
        "properties": {
            "title": {"title": [{"type": "text", "text": {"content": title}}]}
        },
        "children": children,
    }

    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            "https://api.notion.com/v1/pages",
            headers={
                "Authorization": f"Bearer {token}",
                "Notion-Version": "2022-06-28",
                "Content-Type": "application/json",
            },
            json=payload,
        )
        resp.raise_for_status()
        return resp.json().get("url", "")


# ---------------------------------------------------------------------------
# Agent
# ---------------------------------------------------------------------------

today = datetime.now()
date_str = today.strftime("%Y-%m-%d")
day_name = today.strftime("%A")

instructions = f"""
You are a personal productivity assistant that produces a daily digest.

Workflow for {date_str} ({day_name}):
1. Fetch the top Hacker News posts (use fetch_hn_top_stories).
   - Pick the most interesting 5-7.
   - Summarize what makes each one notable.

2. Search the web for restaurant recommendations (use search_web).
   - Aim for highly-rated restaurants in Stockholm with recent reviews.
   - Pick 3-5 worth trying this weekend.

3. Derive 2-4 action items from the news and restaurants.

4. Create a Notion page titled "Daily Digest - {date_str}" (use
   create_notion_page) with sections:
   - "Top News"
   - "Restaurant Picks"
   - "Actions"
   Render each section's body as concise plain text — one line per item with
   the title and link.

Be concise. Always include source URLs. Iterate through every step; do not
return an empty digest.
"""

hooks = Hooks(
    on_iteration_start=lambda ctx: print(f"\n--- Iteration {ctx['iteration']} ---"),
    on_tool_start=lambda ctx: print(
        f"  -> {ctx['name']}({json.dumps(ctx['input'])[:120]}"
        f"{'...' if len(json.dumps(ctx['input'])) > 120 else ''})"
    ),
    on_tool_end=lambda ctx: (
        print(f"  x {ctx['name']} failed ({ctx['duration_ms']:.0f}ms): {ctx['error']}")
        if ctx.get("error")
        else print(
            f"  <- {ctx['name']} ({ctx['duration_ms']:.0f}ms): "
            f"{json.dumps(ctx['output'])[:150]}"
            f"{'...' if len(json.dumps(ctx['output'])) > 150 else ''}"
        )
    ),
    on_agent_end=lambda ctx: (
        print(f"\nx Agent failed: {ctx['error']}")
        if ctx.get("error")
        else print(f"\nv Agent completed in {ctx['result'].meta.iterations} iteration(s)")
        if ctx.get("result")
        else None
    ),
)

agent = Agent(
    name="daily-digest",
    instructions=instructions,
    output_schema=DailyDigest,
    model="anthropic/claude-sonnet-4-6",
    tools=[fetch_hn_top_stories, search_web, create_notion_page],
    max_iterations=15,
    hooks=hooks,
)


# ---------------------------------------------------------------------------
# Run
# ---------------------------------------------------------------------------


async def main() -> None:
    for key in ("OPPER_API_KEY", "JINA_API_KEY", "NOTION_TOKEN", "NOTION_PARENT_PAGE_ID"):
        if not os.environ.get(key):
            raise SystemExit(
                f"Missing required env var: {key}\n"
                "See the docstring at the top of this file for setup steps."
            )

    print("Daily Digest Agent")
    print("=" * 60)
    print(f"Date: {date_str} ({day_name})\n")
    print("Running...\n")

    try:
        result = await agent.run(
            f"Create my daily digest for {date_str} ({day_name})."
        )
        digest = result.output
        print("=" * 60)
        print("DAILY DIGEST CREATED")
        print("=" * 60)
        print(f"\nDate: {digest.date}")
        print(f"News items: {len(digest.news)}")
        print(f"Restaurants: {len(digest.restaurants)}")
        print(f"Actions: {len(digest.actions)}")
        print(f"\nNotion page: {digest.notion_page_url}")
        print(f"Tokens used: {result.meta.usage.total_tokens}")
        print(f"Iterations: {result.meta.iterations}")
        print(f"Tool calls: {len(result.meta.tool_calls)}")
    except Exception as err:
        print(f"\nError: {err}")
        cause = getattr(err, "cause", None) or getattr(err, "__cause__", None)
        if cause:
            print(f"Cause: {cause}")
        raise SystemExit(1)


if __name__ == "__main__":
    asyncio.run(main())
